[curl] Add HTTP/3 enabling option and settings

kshukuwa opened the pull request:
#### 7327aab70ea504499e0f1a14c78f55863233afd9
<pre>
[curl] Add HTTP/3 enabling option and settings
<a href="https://bugs.webkit.org/show_bug.cgi?id=255978">https://bugs.webkit.org/show_bug.cgi?id=255978</a>

Reviewed by Don Olmstead and Fujii Hironori.

Add HTTP/3 enabling option and settings to make it easier
to check HTTP/3 behavior.

* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlContext::CurlContext):
(WebCore::CurlHandle::enableHttp):
(WebCore::CurlHandle::enableAltSvc):
(WebCore::CurlContext::isAltSvcEnabled const): Deleted.
(WebCore::CurlContext::isHttp2Enabled const): Deleted.
* Source/WebCore/platform/network/curl/CurlContext.h:
(WebCore::CurlContext::isAltSvcEnabled const):
(WebCore::CurlContext::isHttp2Enabled const):
(WebCore::CurlContext::isHttp3Enabled const):
* Source/WebCore/platform/network/curl/ResourceResponseCurl.cpp:
(WebCore::ResourceResponse::ResourceResponse):

Canonical link: <a href="https://commits.webkit.org/263447@main">https://commits.webkit.org/263447@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/278e2080410fe15e6ed9ab2dda9e800a98a25ea0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4658 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6155 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4805 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4648 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4740 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5042 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4816 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/4172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6164 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2304 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/4164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9143 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/4164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5784 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4634 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3769 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4159 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8205 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/530 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/4519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont reviewed (approved):
Since we have to explicitly build in support to curl for HTTP/3 I feel like we don’t need an enable flag. @fujii you have thoughts on that?

Everything else looks good to me

fujii commented on Source/cmake/WebKitFeatures.cmake:
```diff
@@ -152,6 +152,7 @@ macro(WEBKIT_OPTION_BEGIN)
     WEBKIT_OPTION_DEFINE(ENABLE_CSS_DEVICE_ADAPTATION "Toggle CSS Device Adaptation support" PRIVATE OFF)
     WEBKIT_OPTION_DEFINE(ENABLE_CSS_IMAGE_RESOLUTION "Toggle CSS image-resolution support" PRIVATE OFF)
     WEBKIT_OPTION_DEFINE(ENABLE_CSS_PAINTING_API "Toggle CSS Painting API support" PRIVATE OFF)
+    WEBKIT_OPTION_DEFINE(ENABLE_CURL_HTTP3 "Toggle HTTP/3 support" PRIVATE OFF)
```
It doesn't seem like a good idea that adding curl specific option in WebKitFeatures.cmake.
OptionsGTK.cmake and OptionsWPE.cmake contains port specific options.

I don't object to add `ENABLE_CURL_HTTP3` if this is useful for your development.
But, I think you are the only one who will use `ENABLE_CURL_HTTP3`.
You can add compile-time flag in CurlContext.cpp. You won't have to recompile whole WebKit just by changing it.

I think WinCairo should always enable HTTP/3 and Alt-Svc if they are mature enough.
Do you want to disable HTTP/3 for PlayStation port at compile-time even after WinCairo enables it?

fujii commented on Source/WebCore/platform/network/curl/CurlContext.cpp:
```diff
@@ -478,8 +488,11 @@ void CurlHandle::enableRequestHeaders()
-    if (m_url.protocolIs("https"_s) && CurlContext::singleton().isHttp2Enabled()) {
-        curl_easy_setopt(m_handle, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2TLS);
+    auto isHttp2Enabled = CurlContext::singleton().isHttp2Enabled();
+    auto isHttp3Enabled = CurlContext::singleton().isHttp3Enabled() && CurlContext::singleton().isAltSvcEnabled();
```
If Alt-Svc is necessary for HTTP/3, I think this condition "&& CurlContext::singleton().isAltSvcEnabled()" should be moved into `CurlContext::isHttp3Enable`.

kshukuwa committed:
[curl] Add HTTP/3 enabling option and settings
https://bugs.webkit.org/show_bug.cgi?id=255978

Reviewed by Don Olmstead and Fujii Hironori.

Add HTTP/3 enabling option and settings to make it easier
to check HTTP/3 behavior.

* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlContext::CurlContext):
(WebCore::CurlHandle::enableHttp):
(WebCore::CurlHandle::enableAltSvc):
(WebCore::CurlContext::isAltSvcEnabled const): Deleted.
(WebCore::CurlContext::isHttp2Enabled const): Deleted.
* Source/WebCore/platform/network/curl/CurlContext.h:
(WebCore::CurlContext::isAltSvcEnabled const):
(WebCore::CurlContext::isHttp2Enabled const):
(WebCore::CurlContext::isHttp3Enabled const):
* Source/WebCore/platform/network/curl/ResourceResponseCurl.cpp:
(WebCore::ResourceResponse::ResourceResponse):

Canonical link: https://commits.webkit.org/263447@main
